### PR TITLE
Validation during training and other goodies

### LINKED ISSUE #5: Validation set accuracy every k epochs
We can either do this, or split the training set into something 95%-5%, and only used validation for occasional validation. 

What do you guys think?

### PR DESCRIPTION
**Pull request by AndreiBarsan**
PLEASE REVIEW CAREFULLY; mistakes here may lead to subtle errors in error estimation.

Changes:
 * Main (fixes #5 in a more immediately-useful way): 5% of training data is now held out, and used to estimate validation error a few times an epoch (configurable). The 5% is also configurable, but it's ~10k samples, which IMHO is more than enough to tell if we're overfitting or screwing up in any other way.
 * Improved `fabfile.py` by removing hardcoded paths. Default path settings can still be manually overridden in most commands if necessary.
 * Fabric task for downloading shit: `fab get_data`
 * Can now parameterize LSTM module much better. Includes option (untested yet) for bidirectional LSTMs. The number and size of LSTM layers can now be set on the command line.
 * Fixed random seed in `trainMLP.py` (partial fix for #8; still need same in evaluation code)

### CODE REVIEW

File: visualqa/trainMLP.py [hunk 1]
```diff
@@ -38,10 +42,14 @@ def load_train_data(data_root):
...
-    parser.add_argument('-num_hidden_layers', type=int, default=3)
+    parser.add_argument('-num_hidden_units', type=int, default=1024,
+                        help="Hidden layer size in the final MLP.")
+    parser.add_argument('-num_hidden_layers', type=int, default=3,
+                        help="Number of hidden layers in the final MLP.")
+    parser.add_argument('-lstm_num_layers', type=int, default=1)
+    parser.add_argument('-lstm_layer_size', type=int, default=256)
     parser.add_argument('-dropout', type=float, default=0.5)
-    parser.add_argument('-activation', type=str, default='tanh')
+    parser.add_argument('-activation', type=str, default='relu')
```
Comment (bernhard2202) [inline visualqa/trainMLP.py]: good point I didn't see that.

File: visualqa/trainMLP.py [hunk 2]
```diff
@@ -95,33 +106,81 @@ def main():
     data_root = args.dataroot
     experiment_root = args.experiment_root
 
+    # Set both the numpy and the Python random seeds.
+    random.seed(RANDOM_SEED)
+    np.random.seed(RANDOM_SEED)
```
Comment (bernhard2202) [inline visualqa/trainMLP.py]: How does weight initialization in Keras work? I guess we also need to set a seed?

File: visualqa/trainMLP.py [hunk 3]
```diff
@@ -95,33 +106,81 @@ def main():
...
+    train_all_count = len(questions_train_all)
+    valid_count = int(train_all_count * args.valid_ratio)
+    train_count = train_all_count - valid_count
+
+    print("We have {0} total Q-A pairs. Will use {1:.2f}% for validation, "
+          "which is {2} data points. {3} data points will be used for "
+          "actual training.".format(train_all_count, args.valid_ratio * 100.0,
+                                    valid_count, train_count))
+
+    questions_train = questions_train_all[:train_count]
```
Comment (bernhard2202) [inline visualqa/trainMLP.py]: Are you sure questions are not ordered in any kind like yes/no before numbers? In that case we would need to permute everything to really select a random subset.

File: visualqa/trainMLP.py (hunk 1 above)
Comment (AndreiBarsan) [inline visualqa/trainMLP.py]: Nah, I think it was inherited from the original codebase. I don't expect it to make that big of a difference, but I guess relu is better as a rule of thumb.

File: visualqa/trainMLP.py (hunk 3 above)
Comment (AndreiBarsan) [inline visualqa/trainMLP.py]: They may be, yes. But I'm shuffling everything just above, so it should be OK.

File: visualqa/trainMLP.py (hunk 2 above)
Comment (AndreiBarsan) [inline visualqa/trainMLP.py]: Great point; I checked it out. In principle it relies on the numpy seed and (I think) the TF seed, so I'll try setting that too.
